src/transport/SecureSessionTable.cpp

```diff
@@ -0,0 +1,277 @@
+#include "lib/support/ScopedBuffer.h"
+#include <access/AuthMode.h>
+#include <transport/SecureSession.h>
+#include <transport/SecureSessionTable.h>
+
+namespace chip {
+namespace Transport {
+
+Optional<SessionHandle> SecureSessionTable::CreateNewSecureSessionForTest(SecureSession::Type secureSessionType,
+                                                                          uint16_t localSessionId, NodeId localNodeId,
+                                                                          NodeId peerNodeId, CATValues peerCATs,
+                                                                          uint16_t peerSessionId, FabricIndex fabricIndex,
+                                                                          const ReliableMessageProtocolConfig & config)
+{
+    if (secureSessionType == SecureSession::Type::kCASE)
+    {
+        if ((fabricIndex == kUndefinedFabricIndex) || (localNodeId == kUndefinedNodeId) || (peerNodeId == kUndefinedNodeId))
+        {
+            return Optional<SessionHandle>::Missing();
+        }
+    }
+    else if (secureSessionType == SecureSession::Type::kPASE)
+    {
+        if ((fabricIndex != kUndefinedFabricIndex) || (localNodeId != kUndefinedNodeId) || (peerNodeId != kUndefinedNodeId))
+        {
+            // TODO: This secure session type is infeasible! We must fix the tests
+            if (false)
+            {
+                return Optional<SessionHandle>::Missing();
+            }
+
+            (void) fabricIndex;
+        }
+    }
+
+    SecureSession * result = mEntries.CreateObject(*this, secureSessionType, localSessionId, localNodeId, peerNodeId, peerCATs,
+                                                   peerSessionId, fabricIndex, config);
+    return result != nullptr ? MakeOptional<SessionHandle>(*result) : Optional<SessionHandle>::Missing();
+}
+
+Optional<SessionHandle> SecureSessionTable::CreateNewSecureSession(SecureSession::Type secureSessionType,
+                                                                   ScopedNodeId sessionEvictionHint)
+{
+    Optional<SessionHandle> rv = Optional<SessionHandle>::Missing();
+    SecureSession * allocated  = nullptr;
+
+    auto sessionId = FindUnusedSessionId();
+    VerifyOrExit(sessionId.HasValue(), rv = Optional<SessionHandle>::Missing());
+
+    //
+    // We allocate a new session out of the pool if we have space in it. If we don't, we need
+    // to run the eviction algorithm to get a free slot. We shall ALWAYS be guaranteed to evict
+    // an existing session in the table in normal operating circumstances.
+    //
+    if (mEntries.Allocated() < CHIP_CONFIG_SECURE_SESSION_POOL_SIZE)
+    {
+        allocated = mEntries.CreateObject(*this, secureSessionType, sessionId.Value());
+        VerifyOrExit(allocated != nullptr, rv = Optional<SessionHandle>::Missing());
+    }
+    else
+    {
+        allocated = EvictAndAllocate(sessionId.Value(), secureSessionType, sessionEvictionHint);
```

Comment:
Looks like `EvictAndAllocate` can return a nullptr which you would want to properly error out here.

I think you can move `VerifyOrExit(allocated != nullptr, rv = Optional<SessionHandle>::Missing());` from above to after the if/else statement